node/src/components/consensus/utils.rs

```diff
@@ -52,5 +58,64 @@ pub(crate) fn check_sufficient_finality_signatures(
         });
     }
 
+    // Verify: the total weight decreased by the smallest weight of a single signature should be
+    // below threshold (anti-spam countermeasure).
+    if weight_minus_minimum * U512::from(*lower_bound.denom())
+        > total_weight * U512::from(*lower_bound.numer())
+    {
+        return Err(FinalitySignatureError::TooManySignatures {
+            trusted_validator_weights: trusted_validator_weights.clone(),
+            block_signatures: Box::new(block_signatures.clone()),
+            signature_weight: Box::new(signature_weight),
+            weight_minus_minimum: Box::new(weight_minus_minimum),
+            total_validator_weight: Box::new(total_weight),
+            finality_threshold_fraction,
+        });
+    }
+
     Ok(())
 }
+
+pub(crate) fn get_minimal_set_of_signatures(
+    trusted_validator_weights: &BTreeMap<PublicKey, U512>,
+    finality_threshold_fraction: Ratio<u64>,
+    mut block_signatures: BlockSignatures,
+) -> BlockSignatures {
+    // Calculate the values for comparison.
+    let total_weight: U512 = trusted_validator_weights
+        .iter()
+        .map(|(_, weight)| *weight)
+        .sum();
+
+    let lower_bound = (finality_threshold_fraction + 1) / 2;
+
+    let mut sig_weights: Vec<_> = block_signatures
+        .proofs
+        .keys()
+        .filter_map(|pub_key| {
+            trusted_validator_weights
+                .get(pub_key)
+                .map(|weight| (pub_key.clone(), *weight))
+        })
+        .collect();
+    // Sort descending by weight.
+    sig_weights.sort_unstable_by(|a, b| b.1.cmp(&a.1));
+
+    let mut accumulated_weight = U512::zero();
+    let keys_to_retain: HashSet<_> = sig_weights
+        .into_iter()
+        .take_while(|(_, weight)| {
+            let to_take = accumulated_weight * U512::from(*lower_bound.denom())
+                <= total_weight * U512::from(*lower_bound.numer());
+            accumulated_weight += *weight;
+            to_take
```

Comment:
Wouldn't
```rust
take_while(|(_, weight)| {
  accumulated_weight += *weight;
  accumulated_weight >= lower_bound
}
```
Since weights are already in descendending ordered we should stop as soon as we have the minimal set of signatures.

Reply:
If we did that, we'd stop one value too early, wouldn't we? We'd stop once the accumulated weight is over the threshold, but we'd return `false` for the value that brought it over the threshold and effectively we'd collect enough signatures to be just below it.

Reply:
Oh right, I initially wrote `acumulated_weight < lower_bound` but you're correct - we would _not_ take the last necessery element.